### ISSUE TITLE
Flag to type 'catch' variables as 'unknown'.

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#41016]
In TypeScript 4.0, we allowed users to to annotate `catch` variables with `: unknown`; however, it'd be nice if we could have that be the default.

I could imagine a flag like `--useUnknownInCatchVariables` to switch the default type to be `unknown`.

```ts
// @useUnknownInCatchVariables: true

try {
  // ...
}
catch (e) {
  e.toUpperCase(); // error
}
```

_____

Alternatively, I could also imagine a much broader flag that also types parameters as `unknown` as well.

**Comment by MartinJohns** [comment]
@DanielRosenwasser You might want to proofread the title.

**Comment by DanielRosenwasser** [comment]
GitHub seems to be having UI glitches for me or something....

**Comment by AviVahl** [comment]
Default in tsc --init please? :)

**Comment by treybrisbane** [comment]
@DanielRosenwasser will this flag be included in `strict`?

**Comment by gerardolima** [comment]
I believe this should be the default behavior, actually. The inferred type for `err` in `catch (err) { }` should be `unknown`, which it _is_, instead of `any` which is unsafe and highly discouraged.

`unknown` is safer and provides better semantics; I really don't understand why this is not the default for tsc.

**Comment by MartinJohns** [comment]
@gerardolima TypeScript **usually** does not introduce breaking changes without a very good reason, which I'd argue this is not. It's in the same realm with the strict compiler flags, which all are breaking changes and are opt-in.

**Comment by gerardolima** [comment]
Only code that rely on wrong assumptions (and `any`) would brake, @MartinJohns. As extensively demonstrated, exceptions in JS can be just anything and that's why no assumption should be valid on their types by default. This semantic is aligned with the TS type `unknown`, instead of `any`. Obviously, this is my view on the subject and that may not be aligned with TS team.

**Comment by DanielRosenwasser** [comment]
So through [a personal Twitter survey](https://twitter.com/drosenwasser/status/1390054871122944002), current results show that many people have voted for `unknown` as their preference in `catch` clause variables. There's definitely a bias there, but the preference between `any` and `unknown` skews higher towards `unknown` than I would've expected.

That said, there are two things worth mentioning:

* I don't know if people really understand the pain that would come with `unknown` in `catch` variables.
* A lot of people have specified a preference for `Error` or some form of checked exceptions, which is slightly out of scope.

**Comment by samhh** [comment]
I suspect you could broadly split TypeScript devs into two camps, those who're migrating from JavaScript or otherwise just want a barely-typed experience, and those who want their type system to be as safe as possible thus want to avoid `any`. For example, I'd be surprised if anyone wanted `unknown` catch clause variables but didn't also want `JSON.parse` to return `unknown`.

Totally anecdotal, but in codebases I've worked on where we've made use of try/catch for control flow we seldom ever treated it as anything other than `unknown`, it was generally just a value to propagate, be that to the console, an external logging service, etc. Even with a flag like this enabled the exceptions to the rule could easily be overridden with assertions.

**Comment by ritschwumm** [comment]
i suspect the camp migrating from javascript is on the decline - at least in my experience nobody starts a project in plain JS any more when it's possible to use TS instead.

**Comment by MartinJohns** [comment]
@ritschwumm There are **many** projects out there that don't support strict type checks, and new ones are still created each day. Two examples I recently had to suffer from are Angular (which just recently started supporting `strictNullChecks` (released Sep 2016), disabled by default) and NestJS (which does not support strict type checks).

So I wouldn't say the only issue are projects migrating from JS.

**Comment by gerardolima** [comment]
@samhh the following code fails with @typescript-eslint rules due to no-unsafe-assignment; that's what I want to avoid. Variables typed as `any` are obviously handy, but they are _unsafe_ and removing that liability is the whole point of TypeScript. So, _in my opinion_, the behaviour that infers types as `any` should be the optional ones.

```ts
try { throw new Error('any-error') } catch(err) { console.log({ err }) }
```

**Comment by phiresky** [comment]
Just want to mention there's a typescript-eslint rule to disallow inferring of the catch type (to enforce using `catch(err: unknown)`): [no-implicit-any-catch](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-implicit-any-catch.md).

Would be great to have it integrated in typescript though.
